Phil/cp http rebased

psFried opened the pull request:
**Description:**

Resolves #1807 
See #1808 for discussion related to the design.
Introduces a public control plane API, along with the first endpoint for fetching status information. This involved moving a bunch of types from the `agent` crate into the `models` crate, which I've isolated in the first commit.

Also introduces the `flowctl catalog status` subcommand for fetching status information using flowctl.

Best to review commits individually for this one.

**Workflow steps:**

- The API docs can be seen at (http://localhost:8675/api/v1/docs)
- `flowctl catalog status my/task my/collection` will print status information to the terminal. Using `-o json|yaml` will include more detailed information.

**Documentation links affected:**

None yet. At some point, we'll probably want to bless the new agent API as being officially supported, at which point we ought to have a docs page on it. But for now, it's still a baby.

**Notes for reviewers:**

We're still working on consuming this API from the UI, so its still possible we may wish to change this. But given how many files I needed to touch in order to move the status into the `models` crate, I'd prefer to get this reviewed and merged before needing to rebase more.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/estuary/flow/1819)
<!-- Reviewable:end -->

psFried committed:
models: move Status types into models crate

This sets up a future commit to expose status via an http endpoint. We'd like
to have all types that are used in the public API organized under the `models`
crate to make it easy to share code between the client and server. So this
moves the status types and refactors controllers to use functions that take an
explicit `status` parameter instead of `self`.  There are no behavior changes
here.

This also impacted publications, discovers, and evolutions, since we expose a
few types from those as part of the status. I moved over only the minimal set
of types that are embedded within a status, though it may be necessary to move
more later as we introduce http endpoints for discover and publish operations.

psFried committed:
agent: introduce status HTTP API and OpenAPI docs

Adds the first public HTTP endpoint to agent, for fetching the status for a
list of live specs.  This includes some boilerplate for generating an OpenAPI
spec from our axum handlers, and API docs that are rendered from that spec
using Scalar.

I tried to establish some repeatable patterns for writing API handlers with
generated documentation and consistent error handling. The objective was to
make it pretty straight forward to add more endpoints. Part of this was
introducing a separation between the "public" APIs that we wish to make
available to all clients, and the private APIs for which we don't want to
guarantee backward compatibility. The idea is to only add apis to the `public`
module when we want to support it for use by end users (which implies at least
some level of backward compatibility). Only the public APIs are covered by the
OpenAPI spec.

The `aide` crate is used for generating the OpenAPI spec. It's not _great_,
but seems to be better than the alternative, `utoipa`, which requires
macros for defining routes and isn't directly compatible with `schemars`.

psFried committed:
flowctl: add `catalog status` subcommand

Adds `flowctl catalog status`, which uses the new status API to fetch the
current status of one or more live specs and print them to the terminal.

jshearer reviewed (approved):
I spot checked a bunch of code that moved around in the refactor commit, and it all looked unchanged as you said it was. I didn't exhaustively check everything.

Everything else LGTM, I just had a couple questions and nits. Will be great to get this out there!

psFried commented on crates/agent/src/api/public/status.rs:
```diff
@@ -0,0 +1,80 @@
+#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
+pub struct StatusQuery {
+    /// The catalog name of the live spec to get the status of
+    pub name: Vec<String>,
```
The way to request multiple statuses would be `?name=foo&name=bar`, which is why the name is singular. This kinda seems like the best way to do it, since otherwise we'd need to somehow encode multiple values in one parameter, and we'd _still_ need to deal with the possibility that the request might provide a query parameter multiple times. IMHO just providing the query parameter multiple times is the most straight forward way to do this, though I'd want to hear if you disagree.

jshearer replied:
Yeah IMO `?name=foo&name=bar` is perfectly fine, and agreed `name` names sense in that context

psFried commented on crates/models/src/status/mod.rs:
```diff
@@ -4,10 +4,51 @@ pub mod collection;
+    pub last_pub_id: Id,
+    /// The id of the last successful publication of the spec, regardless of
+    /// whether the spec was modified. This value can be compared against the
+    /// value of `/controller_status/activations/last_activated` in order to
```
Ahh, that should say `/status/activations/last_activated`

psFried committed:
flowctl: show activation status in status table

Adds a "Activation Complete" column to the status table output, which shows
whether the spec has been successfully activated in the data plane.

psFried commented on crates/flowctl/src/catalog/status.rs:
```diff
@@ -0,0 +1,57 @@
+    type CellValue = JsonCell;
+
+    fn table_headers(_alt: Self::TableAlt) -> Vec<&'static str> {
+        vec![
```
I just added another commit to show that 👍
